src/cpp/session/modules/SessionProfiler.cpp

```diff
@@ -65,10 +65,40 @@ void handleProfilerResReq(const http::Request& request,
    pResponse->setCacheableFile(profileResource, request);
 }
 
+void onDocPendingRemove(const std::string &id)
+{
+   // retrieve document from source database
+   boost::shared_ptr<source_database::SourceDocument> pDoc(
+               new source_database::SourceDocument());
+   Error error = source_database::get(id, pDoc);
```

Comment:
Just a suggestion: it looks like all the listeners to `onDocPendingRemove` pull the document from the source database 
(via `::get`). This is not super cheap (IIRC it reads the doc properties *and* all its contents from disk). It might be worth modifying `onDocPendingRemove` so that it extracts the source document once and passes it as an argument to the event.